File: app/controllers/integration/ExamRecordAPIController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.integration;

import be.objectify.deadbolt.java.actions.SubjectNotPresent;
import controllers.base.BaseController;
import io.ebean.DB;
import java.util.List;
import models.ExamRecord;
import models.dto.ExamScore;
import org.joda.time.DateTime;
import org.joda.time.format.ISODateTimeFormat;
import play.libs.Json;
import play.mvc.Result;

public class ExamRecordAPIController extends BaseController {

    @SubjectNotPresent
    public Result getNewRecords(String startDate) {
        return ok(Json.toJson(getScores(startDate)));
    }

    private List<ExamScore> getScores(String startDate) {
        DateTime start = ISODateTimeFormat.dateTimeParser().parseDateTime(startDate);
        return DB
            .find(ExamRecord.class)
            .fetch("examScore")
            .where()
            .gt("timeStamp", start)
            .findList()
            .stream()
            .map(ExamRecord::getExamScore)
            .toList();
    }
}

File: app/controllers/integration/ExamRecordAPIController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.integration

import io.ebean.DB
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.admin.ExamScore
import models.assessment.ExamRecord
import org.joda.time.format.ISODateTimeFormat
import play.api.mvc.{Action, AnyContent, BaseController, ControllerComponents}
import security.scala.Auth.subjectNotPresent

import javax.inject.Inject
import scala.concurrent.ExecutionContext

class ExamRecordAPIController @Inject() (
    val controllerComponents: ControllerComponents,
    implicit val ec: ExecutionContext
) extends BaseController
    with DbApiHelper
    with JavaApiHelper:

  def getNewRecords(startDate: String): Action[AnyContent] =
    Action.andThen(subjectNotPresent) { _ =>
      val scores = getScores(startDate)
      Ok(scores.asJson)
    }

  private def getScores(startDate: String): List[ExamScore] =
    val start = ISODateTimeFormat.dateTimeParser().parseDateTime(startDate)
    DB.find(classOf[ExamRecord])
      .fetch("examScore")
      .where()
      .eq("releasable", true)
      .gt("timeStamp", start)
      .list
      .flatMap(record => Option(record.getExamScore))
